Unify THAllocator and THCDeviceAllocator with at::Allocator

ezyang opened the pull request:
Stacked on #9087

```
    Unify THAllocator and THCDeviceAllocator with at::Allocator
    
    - at::Allocator is the type that "survived" this refactor, but
      it grew a void* ctx argument.  This brings it in line
      with the THAllocator interface.  Arguably, one might say that
      void* ctx is bad design.  But...
    
        - We definitely need some sort of context for deleters,
          because sometimes we need more information than the data
          pointer itself to properly deallocate.  A simple example
          is a dlpack tensor; to free the data, we need an
          enclosing struct which contains the actual function pointer
          to perform deletion.
    
        - So, it's quite plausible we can pack the deleter together
          with its context data (e.g., an object with a function
          pointer and its context).  This is not a bad idea, but this
          patch is pretty long already, so we leave it for later.
    
    - THCDeviceAllocator has a few differences from at::Allocator:
    
        - It used to return a cudaError_t.  Now, allocators
          are expected to check the error status immediately and throw
          an exception if there was an error.  It turns out that this
          is what was immediately done after all occurrences of
          allocate/release, so it wasn't a big deal (although some
          subsidiary interfaces had to themselves be converted to
          not return cudaError_t).
    
          There is one notable exception to this, and it is how
          we handle CUDA OOM: if this occurs, we attempt to return
          unused memory to the system and try again.  This is now
          handled by a catch-all try-catch block.
    
        - It used to take the CUDA stream to perform the allocation
          on as an argument.  However, it turned out that all call
          sites, this stream was the stream for the current device.
          So we can push this into the allocator (and the choice,
          in the future, could be made explicitly by twiddling
          thread local state.)
    
        - It held two extra methods, emptyCache and cacheInfo, specifically
          for interacting with some state in THCCachingAllocator.
          But this "generality" was a lie, since THCCachingAllocator
          was the only allocator that actually implemented these
          methods, and there is actually a bunch of code in THC
          which assumes that it is the caching allocator that is
          the underlying allocator for CUDA allocations.  So I
          folded these two methods into this interface as
          THCCachingAllocator_emptyCache and THCCachingAllocator_cacheInfo.
    
        - It held its context directly inside the THCDeviceAllocator
          struct.  This context has been moved out into whatever
          is holding the at::Allocator*.

    - I took the opportunity to replace all global variables with
      getter functions; e.g., where you had &THCudaHostAllocator,
      you now have getTHCudaHostAllocator()
    
    - Realloc is no more.  (I'm not sure how important it was for
      the CPU case.  However, if it was important, we have to add
      it back very carefully. There is no way to write a "default"
      implementation of realloc using malloc and free, because some
      mechanism of copying data from the old region to the new
      region is needed, and "how to copy data" is definitely not
      in the purview of an allocator.  Furthermore, implementations
      of realloc usually have side conditions: they usually only
      work on plain-old-data, because the reallocator doesn't know
      what the type of the data it is copying are.)  It's a lot
      simpler for the design if we give up the slight optimization
      we get in the CPU case from deferring the reallocation to the
      system malloc() implementation, so I've dropped it for now.
    
    - allocatorVoidPtr is no more!  Now it's just an at::Allocator*
    
    I tried not to make big semantic changes to any of the downstream
    allocators in this patch (as it was getting big), but a lot of the
    downstream allocators are designed a bit strangely, since they
    were built under the assumption that TH/THC could not be changed.
    They definitely merit refactoring.
   
```

ezyang committed:
Unify THAllocator and THCDeviceAllocator with at::Allocator

- at::Allocator is the type that "survived" this refactor, but
  it grew a void* ctx argument.  This brings it in line
  with the THAllocator interface.  Arguably, one might say that
  void* ctx is bad design.  But...

    - We definitely need some sort of context for deleters,
      because sometimes we need more information than the data
      pointer itself to properly deallocate.  A simple example
      is a dlpack tensor; to free the data, we need an
      enclosing struct which contains the actual function pointer
      to perform deletion.

    - So, it's quite plausible we can pack the deleter together
      with its context data (e.g., an object with a function
      pointer and its context).  This is not a bad idea, but this
      patch is pretty long already, so we leave it for later.

- THCDeviceAllocator has a few differences from at::Allocator:

    - It used to return a cudaError_t.  Now, allocators
      are expected to check the error status immediately and throw
      an exception if there was an error.  It turns out that this
      is what was immediately done after all occurrences of
      allocate/release, so it wasn't a big deal (although some
      subsidiary interfaces had to themselves be converted to
      not return cudaError_t).

      There is one notable exception to this, and it is how
      we handle CUDA OOM: if this occurs, we attempt to return
      unused memory to the system and try again.  This is now
      handled by a catch-all try-catch block.

    - It used to take the CUDA stream to perform the allocation
      on as an argument.  However, it turned out that all call
      sites, this stream was the stream for the current device.
      So we can push this into the allocator (and the choice,
      in the future, could be made explicitly by twiddling
      thread local state.)

    - It held two extra methods, emptyCache and cacheInfo, specifically
      for interacting with some state in THCCachingAllocator.
      But this "generality" was a lie, since THCCachingAllocator
      was the only allocator that actually implemented these
      methods, and there is actually a bunch of code in THC
      which assumes that it is the caching allocator that is
      the underlying allocator for CUDA allocations.  So I
      folded these two methods into this interface as
      THCCachingAllocator_emptyCache and THCCachingAllocator_cacheInfo.

    - It held its context directly inside the THCDeviceAllocator
      struct.  This context has been moved out into whatever
      is holding the at::Allocator*.

- I took the opportunity to replace all global variables with
  getter functions; e.g., where you had &THCudaHostAllocator,
  you now have getTHCudaHostAllocator()

- Realloc is no more.  (I'm not sure how important it was for
  the CPU case.  However, if it was important, we have to add
  it back very carefully. There is no way to write a "default"
  implementation of realloc using malloc and free, because some
  mechanism of copying data from the old region to the new
  region is needed, and "how to copy data" is definitely not
  in the purview of an allocator.  Furthermore, implementations
  of realloc usually have side conditions: they usually only
  work on plain-old-data, because the reallocator doesn't know
  what the type of the data it is copying are.)  It's a lot
  simpler for the design if we give up the slight optimization
  we get in the CPU case from deferring the reallocation to the
  system malloc() implementation, so I've dropped it for now.

- allocatorVoidPtr is no more!  Now it's just an at::Allocator*

I tried not to make big semantic changes to any of the downstream
allocators in this patch (as it was getting big), but a lot of the
downstream allocators are designed a bit strangely, since they
were built under the assumption that TH/THC could not be changed.
They definitely merit refactoring.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

colesbury reviewed (approved):
Nice, looks good

colesbury reviewed (approved):
Nice, looks good

gchanan commented on aten/src/THC/THCAllocator.cpp:
```diff
@@ -1,80 +1,68 @@
-  err = cudaIpcCloseMemHandle(devPtr);
+struct THCIpcAllocator : public at::Allocator {
+  void* allocate(void* ctx, size_t size) const override {
+    AT_ERROR("THCIpcAllocator.malloc() not supported");
```
should the error message be about allocate instead of malloc?

gchanan commented on aten/src/ATen/Allocator.h:
```diff
@@ -3,14 +3,28 @@
-  virtual void* allocate(size_t n) const = 0;
-  virtual void deallocate(void* ptr) const = 0;
+  virtual void* allocate(void* ctx, size_t n) const = 0;
+  virtual void deallocate(void* ctx, void* ptr) const = 0;
```
do you think there's any place for making a wrapper struct around the context?  An API with two void *s makes me nervous.

ezyang replied:
Yes, I'm working on this in the next patchset

ezyang commented on aten/src/TH/THAllocator.h:
```diff
@@ -11,18 +11,19 @@
-} THAllocator;
+#ifdef __cplusplus
+#include <ATen/Allocator.h>
+using THAllocator = at::Allocator;
```
According to CI, it does work :) (And yes, we need it to be available from C, because a lot of C function headers refer to it).

Actually, I am kind of thinking that we should use `#ifdef __cplusplus` more liberally in the TH headers, as an alternative way of C++ifying headers instead of the `hpp` convention.

gchanan replied:
Oh, I meant does anything from outside of torch (e.g. the CI, internal code) actually use these functions (so they would care if the function decls moved to C++).

ezyang committed:
Add weak pointer and finalizer support directly to THStorage.

The underlying use-case is the file descriptor to storage cache in
torch.multiprocessing.reductions.  Previously, this was implemented by wrapping
an existing allocator with a "weak ref" allocator which also knew to null out
the weak reference when the storage died.  This is terribly oblique, and
prevents us from refactoring the allocators to get rid of per-storage allocator
state.

So instead of going through this fiasco, we instead directly implement weak
pointers and finalizers in THStorage.  Weak pointers to THStorage retain the
THStorage struct, but not the data_ptr.  When all strong references die,
data_ptr dies and the finalizers get invoked.

There is one major hazard in this patch, which is what happens if you
repeatedly call _weak_ref on a storage.  For cleanliness, we no longer
shove our grubby fingers into the finalizer struct to see if there is already
a Python object for the weak reference and return it; we just create a new one
(no one is checking these Python objects for identity).  This means if you
keep calling it, we'll keep piling on finalizers.  That's bad! But I am
not going to fix it until it is actually a problem for someone, because
then we need to add another caching layer.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
This is blocked on a major design problem which means I can't port some external code to work on this PR, so the way I will merge this is to glob this into a bigger redesign that I had planned.
